tinygrad/shape/symbolic.py

```diff
@@ -21,7 +21,11 @@ def render(self, ops=None, ctx=None) -> str:
     return ops[type(self)](self, ops, ctx)
   def vars(self): return []
   # expand a Node into List[Node] that enumerates the underlying Variables from min to max
-  def expand(self) -> List[Node]: raise RuntimeError(self.__class__.__name__)
+  @functools.lru_cache(maxsize=None)
+  def expand(self, ctx:Tuple[Variable, ...]=None) -> List[Node]:
+    if ctx is None: ctx = tuple(v for v in set(self.vars()) if v.expr is None)
+    replacements = itertools.product(*[[NumNode(x) for x in range(v.min, v.max + 1)] for v in ctx[::-1]])
+    return [self.substitute(dict(zip(ctx[::-1], rep))) for rep in replacements]
```

Comment:
Can replacements be empty?

Reply:
Nice catch, looks like python is one step ahead though XD
```
>>> list(itertools.product())
[()]
```

Reply:
Shouldn't this be an empty list? Does adding a fast path after the None check help?
If not ctx: return []

Reply:
replacements is never empty, since expanding a symbolic expression should at least yield 1 symbolic expression